week7/cooper_client_react/01_cooper_calculator.md

```diff
@@ -0,0 +1,543 @@
+        return 'invalid range';
+    }
+  }
+
+  const ageRange = ageRangeTable(parseInt(age));
+
+  if (ageRange === 'invalid range') {
+    return 'Invalid age range';
+  }
+
+  const distanceRanges = cooperTable[gender.toLowerCase()][
+    ageRange
+  ];
+
+  let ratingIndex;
+
+  distanceRanges.forEach((dRange, index) => {
+    if (
+      (dRange.match(/>\d*/) && distance >= parseInt(dRange.slice(1), 10)) ||
+      (dRange.match(/<\d*/) && distance < parseInt(dRange.slice(1), 10))
+    ) {
+      ratingIndex = index;
+    } else {
+      const minMax = dRange.split('-');
+      const min = parseInt(minMax[0], 10);
+      const max = parseInt(minMax[1], 10);
+
+      if (distance >= min && distance <= max) {
+        ratingIndex = index;
+      }
+    }
+  });
+  return ratings[ratingIndex];
+}
+
+export default CooperCalculator;
+```
+
+If you run the component test now, both tests should go green. Let's run the feature tests again, everything should go green right. We have a component that displays the correct result.
+
+![text](cooper_challenge_component_test_green.png)
+
+BUT, as you can see when we run the feature tests, nothing has changed. That's because we haven't actually rendered the `DiplayCooperResult` component in the `App` component. Consequently, we haven't passed in any props to the `DisplayCooperResult` component. We haven't even saved what gets written in the input fields. So, the actual `DiplayCooperResult` component works, but not with the rest of the application. This is the reason they are called component tests.
+
+Let's add some code to the `App` component.
+
+```js
+import React, { Component } from 'react';
+import DisplayCooperResult from './Components/DisplayCooperResult';
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      distance: '',
+      gender: 'female',
+      age: ''
+    }
+  }
+
+  onChange(event) {
+    this.setState({
+      [event.target.id]: event.target.value
+    })
+  }
+
+ render() {
+    return (
+      <div>
+        <div>
+          <label>Distance</label>
+          <input id="distance" onChange={this.onChange.bind(this)}></input>
+        </div>
+
+        <select id="gender" onChange={this.onChange.bind(this)}>
+          <option value="female">Female</option>
+          <option value="male">Male</option>
+        </select>
+
+        <div>
+          <label>Age</label>
+          <input id="age" onChange={this.onChange.bind(this)}></input>
+        </div>
+
+        <DisplayCooperResult
+          distance={this.state.distance}
+          gender={this.state.gender}
+          age={this.state.age}
+        />
+      </div>
+    );
+  }
+}
+
+export default App;
+```
+
+If you run the test now everything should go green!
+
+Now we need to clean up the `App` component, we dont want to have the input fields there. The `App` component should be as clean as possible, that is why we will extract the input fields to their own component.
```

Comment:
When our test now going green, let's refactor our code using the single responsibility principle.

Reply:
With the current state of how we are building this application, mentioning SRP does not feel relevant